### ISSUE TITLE
Discussion: Missing WPF features

### ISSUE THREAD

**Issue by hshristov** [microsoft/microsoft-ui-xaml#3009]
In WPF there is **OnRender** method which could be overriden in order to do custom drawing. There is also **FormattedText** class that can be use to measure and draw text in the **OnRender** method.
These classes are missing from **WinUI 3**.

What would be the recommended way to measure and draw text in **WinUI 3** without using **TextBlock** control?

**Comment by robloo** [comment]
Proper text measuring API is likely needed for https://github.com/microsoft/microsoft-ui-xaml/issues/1520 as well

**Comment by codendone** [comment]
**OnRender:** The rough equivalent of **OnRender** is to create Composition visuals (SpriteVisual/etc.) and attach those to an element via ElementCompositionPreview.SetElementChildVisual(). You'd need to listen for any events which may require you to update those visuals, such as the element size changing.

The Composition APIs are limited on what they can render, though, so this does currently mean that things like text would require app code custom drawing the text into a surface, such as by dropping down to Direct2D and DirectWrite.

**Text measurement and drawing:** Currently our recommendation for text measurement is to create a TextBlock and call Measure() on it. This obviously only works if the measurement you want is only for the complete bounds, and not for more specific measurements, such as the height of one line of wrapping text. And this doesn't help for rendering (other than actually putting the TextBlock in the element tree, of course).

We've talked about having a XAML API to do text measurement, and we've talked about exposing DirectWrite APIs in WinRT, particularly the DWriteTextLayout. Either of these would provide more fine-grained measurement capabilities.
